Fix top-left resizing - account for new position when calling back onResize

conor-kelleher opened the pull request:
## What's wrong?
Hey, love the library. Works really well for the most part but, as a few different issues have mentioned, it has some serious issues when resizing from the left or the top, using anything other than top-left positioning. The following open issues are just some that have documented the issue:

- https://github.com/STRML/react-resizable/issues/131
- https://github.com/STRML/react-resizable/issues/119

I've put together a minimal codesandbox to demonstrate the issue as well: 
https://codesandbox.io/s/react-playground-vwlpk?file=/index.js

Samples from the above codesandbox:
* Dragging to the right: smooth
[![Image from Gyazo](https://i.gyazo.com/547b6702a5910a58dd11d65c6efad055.gif)](https://gyazo.com/547b6702a5910a58dd11d65c6efad055)

* Dragging to the left: jumpy/flickering
[![Image from Gyazo](https://i.gyazo.com/f727608700cdefe106de907809c34c6e.gif)](https://gyazo.com/f727608700cdefe106de907809c34c6e)

## The Cause?
The problem is that on subsequent calls of the `resizeHandler` function, it's reporting a resize based on the `deltaX` and `deltaY` of the movement. However, if calling back through the `onResize` prop results in a position change (for instance, if you are using right-based absolute positioning as in my example above, or if you are manually updating the top/left positions based on the resize), then the drag handle itself might be moved by the re-render, so the next call to `resizeHandler` will report that the handle has moved, even though this is a prop-driven movement rather than a mouse-event driven one.

## The solution
I've added a few lines to track a reference to the drag handle element in the document and in every call of `resizeHandler`, I'm comparing the new top and left of the element on the page with the previous values. A difference in these values will be a result of a positioning change, and so we want to account for this in our `deltaX`/`deltaY` values.

If dragging any handle with an `'n'` in it and its position has changed, then our event's reported `deltaY` value will either be larger or smaller than we expect it to be (because this value will include the position change). Adding this y-position differential to the delta value accounts for this and keeps the callback only reporting on movement that was caused by the mouse event. The same goes for the x-axis and any handle with a `'w'` in it.

## Demo
I've added a new example to the examples demo page which features a new box with omni-directional movement. This can be used to demonstrate a before and after and also shows much more power behind this library.

e.g.
* Old - omni-directional box stuttering dragging top and left
![ezgif-3-9726531015d8](https://user-images.githubusercontent.com/35919832/83949830-dc31f180-a81d-11ea-820a-6c64ff57cb88.gif)

* New - omni-directional box smoothly resizing in all directions
![ezgif-3-8d801f0a91eb](https://user-images.githubusercontent.com/35919832/83949750-6463c700-a81d-11ea-9ea3-4940ea87a478.gif)


Let me know of any issues you come across with it and have a good day!

conor-kelleher commented:
I've just pushed a small fix to scale the amount we're accounting for inversely by the `transformScale` property, since to match the client's screen, we need to counteract the current scale value.

Before:
[![Image from Gyazo](https://i.gyazo.com/14d14ab300067054f4a4e7f44e73b547.gif)](https://gyazo.com/14d14ab300067054f4a4e7f44e73b547)
After:
[![Image from Gyazo](https://i.gyazo.com/b4b72ea5bef6eadeb18e6498c7203271.gif)](https://gyazo.com/b4b72ea5bef6eadeb18e6498c7203271)

STRML commented:
Hi @conor-kelleher. Thanks a lot for this contribution!

I'd like for this to make it into 2.0. #133 contributed a test suite into the application, it would be really great if you could rebase on `master` and add a few cases as part of this PR. If you can, I'll wrap it up in a bow & release as part of 2.0!

STRML commented on lib/Resizable.js:
```diff
@@ -182,6 +186,27 @@ export default class Resizable extends React.Component<Props, State> {
       const canDragY = (this.props.axis === 'both' || this.props.axis === 'y') && ['e', 'w'].indexOf(axis) === -1;
 
+      // Check if handle's position has moved since last resizeHandler call - adjust deltas accordingly
+      if (this.draggingNode == null && e.target instanceof HTMLElement) {
```
Is saving this ref necessary (rather than simply using `e.target`)?

conor-kelleher replied:
I originally had it just using `e.target` on each callback but because the listener is calling back to any movement in the document body, `e.target` will not always be the same value during a drag. It will be whatever element the mouse is currently over. As you can see, using this value directly causes jumpy behaviour when the mouse enters different elements:
[![Image from Gyazo](https://i.gyazo.com/977571a2e3aa46db3d1702a94ee38a27.gif)](https://gyazo.com/977571a2e3aa46db3d1702a94ee38a27),

Hence the need to persist a reference to the same element through multiple callbacks

conor-kelleher replied:
Or an even worse demo:
[![Image from Gyazo](https://i.gyazo.com/14e62780c79298e17990cb785faf0bd2.gif)](https://gyazo.com/14e62780c79298e17990cb785faf0bd2)

STRML commented on lib/Resizable.js:
```diff
@@ -182,6 +186,27 @@ export default class Resizable extends React.Component<Props, State> {
+            deltaY += deltaTopSinceLast / this.props.transformScale;
+          }
+        }
+        this.lastTop = handleRect.top;
```
Can this be a single property (`this.lastHandleRect = handleRect`)?

conor-kelleher replied:
Yep, it can. I've rolled that in now

STRML commented on lib/Resizable.js:
```diff
@@ -182,6 +186,27 @@ export default class Resizable extends React.Component<Props, State> {
+          const deltaLeftSinceLast = handleRect.left - this.lastLeft;
+          const deltaTopSinceLast = handleRect.top - this.lastTop;
+          
+          if (canDragX && axis.indexOf('w') > -1) {
```
The allowable values for axis are:
```js
['s', 'w', 'e', 'n', 'sw', 'nw', 'se', 'ne']
```
So this could be more clear as `axis[axis.length - 1] === 'w'`, similar to https://github.com/STRML/react-resizable/blob/a1fff416339df13b15b23044b026ebbe376a2b9d/lib/Resizable.js#L97.

This of course would be a great use case for `String#endsWith()` but we are trying to avoid prototype methods that [aren't supported in IE](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith).

conor-kelleher replied:
Good point, I've made this change now

STRML commented on lib/Resizable.js:
```diff
@@ -182,6 +186,27 @@ export default class Resizable extends React.Component<Props, State> {
+          if (canDragX && axis.indexOf('w') > -1) {
+            deltaX += deltaLeftSinceLast / this.props.transformScale;
+          }
+          if(canDragY && axis.indexOf('n') > -1) {
```
See above comment, `axis[0] === 'n'` will work here.

conor-kelleher replied:
Very true, much cleaner this way 👌

STRML commented on lib/Resizable.js:
```diff
@@ -182,6 +186,27 @@ export default class Resizable extends React.Component<Props, State> {
       const canDragX = (this.props.axis === 'both' || this.props.axis === 'x') && ['n', 's'].indexOf(axis) === -1;
       const canDragY = (this.props.axis === 'both' || this.props.axis === 'y') && ['e', 'w'].indexOf(axis) === -1;
 
+      // Check if handle's position has moved since last resizeHandler call - adjust deltas accordingly
```
It'd be great if you could expand upon this comment with some of the detail in this PR: that is, what happens, why it only affects `n` and `w` directions, and how this fixes it.

conor-kelleher replied:
Fleshed these comments out a good bit. Hopefully they're clear enough now 🤞

conor-kelleher commented:
I've made the requested changes to the code. Need to add some tests before re-requesting a revie

satendra02 commented:
Hey @conor-kelleher 

Great PR 🎉 . Really looking forward to its release.

conor-kelleher commented:
Hey @STRML, sorry for the delay getting this fixed up. I've made the suggested changes, added a suite of tests for the new functionality and added a new set of examples to the demo, keeping the absolutely positioned ones away from the inline ones:
[![Image from Gyazo](https://i.gyazo.com/87b649be2b0af4ae5dfd4e5b3849e397.gif)](https://gyazo.com/87b649be2b0af4ae5dfd4e5b3849e397)

Just note: after pulling the changes you pushed to master preparing for v2, I couldn't get the example to run since the webpack config was referencing a non-existent file `test/test.js`. I've updated it to point to the new `examples/example.js`. Hope that's right!

STRML commented on lib/Resizable.js:
```diff
@@ -119,7 +119,10 @@ export default class Resizable extends React.Component<Props, ResizableState> {
             deltaY += deltaTopSinceLast / this.props.transformScale;
           }
         }
-        this.lastHandleRect = handleRect;
```
Not that it matters much, but why not just save the whole rect?

conor-kelleher replied:
I hit an issue with flow upon trying to cast an exact object to an inexact type. I'm new to flow so couldnt find a proper way around this, just got it working by taking the exact data I needed. I'm open to changes though

Ashalbulk commented:
I have the same issue. @STRML would this fix available soon?

STRML commented:
This works great. Thanks for the contribution!

vipulbhj commented:
Is there some code sandbox for the final solution ?

vipulbhj commented:
Any plans on updating the official demo to demonstrate this. 

Anyone have any codesandbox for this, I tried the `examples.js`  and it does seem to work for me. Anyone have any examples I can look at ?

gabrielvf1 commented:
Can someone help me? Why when positioning my resize handles on top/left/top-left the behavior is so weird like i show at the GIF? Is something im doing wrong or i need to implement the right behavior by my self?

![ezgif com-gif-maker](https://user-images.githubusercontent.com/26471896/118532196-0f168200-b71d-11eb-8a04-0e30560327ca.gif)
